File: internal/claudefs/mcp.go
package claudefs

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// The MCP servers an account can reach.
//
// They live in <configdir>/.claude.json under "mcpServers", and that file is not
// part of the shared user layer — it cannot be, because it also holds the
// session history and per-project state that make accounts separate in the first
// place. The consequence catches people out: sign a second account in, turn
// sharing on, and its commands, skills and agents all appear, while every MCP
// server you had is silently absent. Nothing says so; the tools are just not
// there, and the agent explains it cannot reach your database.
//
// So the servers are listed per account, and copied between them on request.
//
// Credentials never come back out. A server's env commonly holds a password or
// a token — the ones on this machine hold database passwords — so the listing
// reports which keys exist and never what is in them. Copying moves the values
// from one file to the other without them passing through the API, a log, or a
// command line where every other process could read them.

// MCPServer is one server as the UI needs to see it.
type MCPServer struct {
	Name string `json:"name"`
	// Type is stdio, http, sse or ws. Absent in the file means stdio.
	Type    string   `json:"type"`
	Command string   `json:"command,omitempty"`
	Args    []string `json:"args,omitempty"`
	URL     string   `json:"url,omitempty"`
	// EnvKeys are the names of the environment variables the server is given.
	// The values are deliberately not here: several of them are passwords.
	EnvKeys []string `json:"envKeys,omitempty"`
	// HasSecrets is true when any of those look like a credential, so the UI can
	// say what is about to be copied.
	HasSecrets bool `json:"hasSecrets,omitempty"`
}

// mcpConfigPath is where an account's servers are configured.
func mcpConfigPath(dir string) string { return filepath.Join(dir, ".claude.json") }

// ListMCPServers reads one account's servers.
//
// A missing file is not an error: an account that has never run has no config
// yet, and "none" is the right answer rather than a failure.
func ListMCPServers(dir string) ([]MCPServer, error) {
	raw, err := readMCPConfig(dir)
	if err != nil {
		return nil, err
	}
	servers := rawServers(raw)
	out := make([]MCPServer, 0, len(servers))
	for name, def := range servers {
		out = append(out, describeServer(name, def))
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out, nil
}

func readMCPConfig(dir string) (map[string]json.RawMessage, error) {
	b, err := os.ReadFile(mcpConfigPath(dir))
	if os.IsNotExist(err) {
		return map[string]json.RawMessage{}, nil
	}
	if err != nil {
		return nil, err
	}
	var doc map[string]json.RawMessage
	if err := json.Unmarshal(b, &doc); err != nil {
		return nil, fmt.Errorf("%s is not readable as JSON: %w", mcpConfigPath(dir), err)
	}
	return doc, nil
}

func rawServers(doc map[string]json.RawMessage) map[string]json.RawMessage {
	out := map[string]json.RawMessage{}
	if v, ok := doc["mcpServers"]; ok {
		_ = json.Unmarshal(v, &out)
	}
	return out
}

func describeServer(name string, def json.RawMessage) MCPServer {
	var d struct {
		Type    string            `json:"type"`
		Command string            `json:"command"`
		Args    []string          `json:"args"`
		URL     string            `json:"url"`
		Env     map[string]string `json:"env"`
	}
	_ = json.Unmarshal(def, &d)

	s := MCPServer{Name: name, Type: d.Type, Command: d.Command, Args: d.Args, URL: d.URL}
	if s.Type == "" {
		if d.URL != "" {
			s.Type = "http"
		} else {
			s.Type = "stdio"
		}
	}
	for k := range d.Env {
		s.EnvKeys = append(s.EnvKeys, k)
		if looksSecret(k) {
			s.HasSecrets = true
		}
	}
	sort.Strings(s.EnvKeys)
	return s
}

// looksSecret is a naming heuristic, used only to warn. It decides whether the
// UI says "this carries a credential", never whether a value is protected —
// values are never returned either way.
func looksSecret(key string) bool {
	k := strings.ToLower(key)
	for _, hint := range []string{"password", "passwd", "secret", "token", "key", "pwd", "auth", "credential"} {
		if strings.Contains(k, hint) {
			return true
		}
	}
	return false
}

// CopyMCPServers copies named servers from one account's config into another's,
// and reports which ones were written.
//
// Values and all: this is a file-to-file copy, so a database password goes
// across without being read into the API, written to a log, or passed on a
// command line where anything else on the machine could see it.
//
// Existing servers of the same name are left alone rather than overwritten. A
// name that is already configured is far more likely to be a deliberate variant
// than something wanting replacement, and overwriting one silently would lose a
// working setup.
func CopyMCPServers(fromDir, toDir string, names []string) (copied []string, skipped []string, err error) {
	if sameFile(mcpConfigPath(fromDir), mcpConfigPath(toDir)) {
		return nil, nil, fmt.Errorf("that is the same configuration directory")
	}
	src, err := readMCPConfig(fromDir)
	if err != nil {
		return nil, nil, err
	}
	srcServers := rawServers(src)

	dst, err := readMCPConfig(toDir)
	if err != nil {
		return nil, nil, err
	}
	dstServers := rawServers(dst)

	for _, n := range names {
		def, ok := srcServers[n]
		if !ok {
			skipped = append(skipped, n)
			continue
		}
		if _, exists := dstServers[n]; exists {
			skipped = append(skipped, n)
			continue
		}
		dstServers[n] = def
		copied = append(copied, n)
	}
	if len(copied) == 0 {
		return nil, skipped, nil
	}

	merged, err := json.Marshal(dstServers)
	if err != nil {
		return nil, nil, err
	}
	dst["mcpServers"] = merged
	if err := writeMCPConfig(toDir, dst); err != nil {
		return nil, nil, err
	}
	return copied, skipped, nil
}

// writeMCPConfig rewrites the account's config with everything else in it
// untouched.
//
// The file belongs to the CLI and holds far more than MCP servers — history,
// per-project state, onboarding flags. It is read back as raw JSON and written
// back the same way so that nothing this code does not understand is dropped by
// a round trip through a struct.
//
// A copy of the original is kept beside it first. This is somebody's live
// configuration, and an atomic replace still replaces.
func writeMCPConfig(dir string, doc map[string]json.RawMessage) error {
	path := mcpConfigPath(dir)
	if b, err := os.ReadFile(path); err == nil {
		backup := fmt.Sprintf("%s.backup-%s", path, time.Now().Format("20060102-150405"))
		if err := os.WriteFile(backup, b, 0o600); err != nil {
			return fmt.Errorf("could not back up %s first: %w", path, err)
		}
	}

	out, err := json.MarshalIndent(doc, "", "  ")
	if err != nil {
		return err
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, out, 0o600); err != nil {
		return err
	}
	// Rename over the original, so a reader sees one file or the other and never
	// a half-written one.
	if err := os.Rename(tmp, path); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	return nil
}

func sameFile(a, b string) bool {
	ra, err1 := filepath.Abs(a)
	rb, err2 := filepath.Abs(b)
	if err1 != nil || err2 != nil {
		return a == b
	}
	return strings.EqualFold(ra, rb)
}

File: internal/server/web/mcp.js
/* MCP servers, across every account.
 *
 * The CLI manages these one config directory at a time. With several accounts
 * that is the wrong shape: the question is which of them can reach your
 * database, and the answer is usually "only the first one". `.claude.json` is
 * not part of the shared user layer — it cannot be, it also holds session
 * history and per-project state — so a newly signed-in account has none of your
 * servers, nothing says so, and the agent simply reports that it cannot connect.
 *
 * Values are never shown. A server's env commonly holds a password, so this
 * lists which variables it needs and never what is in them; the copy moves the
 * file's own values across without them passing through here at all.
 */
'use strict';

async function openMCP() {
  const body = el('div', { id: 'mcpBody' }, el('div', { class: 'hint', text: 'Loading…' }));
  modal('MCP servers', body, [['Close', 'btn', closeModal]], true);
  await paintMCP();
}

async function paintMCP() {
  const body = $('#mcpBody');
  if (!body) return;
  let rows = [];
  try {
    rows = await api('/mcp') || [];
  } catch (e) {
    body.innerHTML = '';
    body.append(el('div', { class: 'hint', text: e.message }));
    return;
  }

  body.innerHTML = '';
  body.append(el('div', { class: 'hint', text:
    'Each account has its own set. Sharing your user layer does not cover these — the file they live in also ' +
    'holds session history, which is what keeps accounts separate — so a newly signed-in account starts with none.' }));

  if (!rows.length) {
    body.append(el('div', { class: 'empty', style: 'padding:26px' },
      el('h3', { text: 'No accounts yet' })));
    return;
  }

  const withServers = rows.filter(r => r.servers.length);
  for (const acct of rows) {
    const head = el('div', { class: 'mcp-head' },
      el('strong', { text: acct.name }),
      el('span', { class: 'pill' }, `${acct.servers.length} server${acct.servers.length === 1 ? '' : 's'}`),
      acct.busy ? el('span', { class: 'pill warn', title: 'Its configuration is not written while it is running' }, 'running') : null,
      el('span', { style: 'flex:1' }),
      // Copying into an account only makes sense from one that has some.
      withServers.length && !acct.busy
        ? el('button', { class: 'btn sm', onclick: () => copyInto(acct, withServers) }, 'Copy from…')
        : null);
    body.append(head);

    if (acct.error) {
      body.append(el('div', { class: 'hint', text: acct.error }));
      continue;
    }
    if (!acct.servers.length) {
      body.append(el('div', { class: 'hint', text: 'None. Agents on this account have no MCP tools.' }));
      continue;
    }

    const table = el('table', { class: 'kv mcp-table' });
    for (const s of acct.servers) {
      table.append(el('tr', {},
        el('td', {}, el('span', { class: 'mono', text: s.name })),
        el('td', {}, el('span', { class: 'pill' }, s.type)),
        el('td', { class: 'mcp-cmd mono', title: s.url || [s.command].concat(s.args || []).join(' ') },
          s.url || [s.command].concat(s.args || []).join(' ')),
        el('td', {}, (s.envKeys || []).length
          ? el('span', {
              class: 'pill' + (s.hasSecrets ? ' warn' : ''),
              title: 'Environment it is given: ' + (s.envKeys || []).join(', ') +
                (s.hasSecrets ? '\nOne of these looks like a credential. Its value is never shown here.' : ''),
            }, `${s.envKeys.length} env`)
          : null)));
    }
    body.append(table);
  }
}

// copyInto asks which servers to bring over, and from where.
function copyInto(target, sources) {
  const others = sources.filter(s => s.accountId !== target.accountId);
  if (!others.length) {
    toast('No other account has any to copy', 'bad');
    return;
  }

  const pick = el('select', { id: 'mcpFrom' },
    others.map(o => el('option', { value: o.accountId }, `${o.name} — ${o.servers.length}`)));
  const list = el('div', { id: 'mcpNames', class: 'mcp-pick' });

  const fill = () => {
    const from = others.find(o => o.accountId === pick.value) || others[0];
    list.innerHTML = '';
    for (const s of from.servers) {
      const already = target.servers.some(t => t.name === s.name);
      list.append(el('label', { class: 'switch' },
        el('input', {
          type: 'checkbox', value: s.name,
          checked: already ? null : 'checked',
          disabled: already ? 'disabled' : null,
        }),
        s.name + (already ? ' — already here' : (s.hasSecrets ? ' — carries a credential' : ''))));
    }
  };
  pick.addEventListener('change', fill);
  fill();

  modal(`Copy MCP servers into ${target.name}`, el('div', {},
    el('label', { text: 'From' }), pick,
    el('label', { text: 'Servers' }), list,
    el('div', { class: 'hint', text:
      'The definition is copied with its environment, values included, so the server actually works on the ' +
      'other account. Nothing is overwritten: a name already there is left alone.' })),
    [
      ['Cancel', 'btn', () => { closeModal(); openMCP(); }],
      ['Copy', 'btn primary', async () => {
        const names = [...list.querySelectorAll('input:checked')].map(i => i.value);
        if (!names.length) return toast('Pick at least one', 'bad');
        try {
          const r = await api('/mcp/copy', {
            method: 'POST',
            body: { from: pick.value, to: target.accountId, names },
          });
          const n = (r.copied || []).length;
          toast(n ? `Copied ${n} server${n === 1 ? '' : 's'}` : 'Nothing to copy', n ? 'ok' : 'bad');
        } catch (e) {
          toast(e.message, 'bad');
        }
        closeModal();
        openMCP();
      }],
    ]);
}
